### PR TITLE
Ported some fuel system changes for Thrustmaster quadrant compatibility

### LINKED ISSUE #1630: [BUG] TCA Quadrant Airbus edition ENG 1-2 Turn on levers not work
**Mod Version**
Latest DEV version from the installer.

**Describe the bug**
TCA Quadrant Airbus edition Turn on ENG 1-2  levers not work，Original  A320 is work

**Expected behavior**
ENG 1 and 2 Start lever Turn on ENG start

**Actual behavior**
ENG 1 and 2 Start lever Turn on but not work

Was this working before/when did the issue start occurring?
This time updata add

Is this a problem in the vanilla unmodded game?
Original  A320 is work

### PR DESCRIPTION
**Pull request by Saschl**
Fixes #1630

## Summary of Changes
Patch 1.10.7 changed some ids in the fuel system. This PR aims to adapt those changes by changing the ids wherever they were used. Primary goal was to make the mod compatible with the Thrustmaster Airbus Quadrant fuel switches.

Feel free to close the PR if the changes aren't up to the standards. I just wanted to share what was working for me to get the quadrant running.

## References
Stock Asobo A320 files

Discord username (if different from GitHub): saschl#9432

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by MoparCharger1969** [comment]
I tested this on mine by inserting only the changed files into my latest DEV version (downloaded this AM) and the switches do work correctly now.  Hopefully this will be merged soon so I can keep up with the other dev changes as they continue to improve.  Thanks SaschI!!!

**Comment by Saschl** [comment]
Quick note for the tester(s). The default bindings for the switches are quite buggy (eng. switch 1 activates both swiches ingame, eng. switch 2 makes ingame switch 2 go crazy). And the engine mode knob is not functional. I use the mapping from these posts: [LINK](https://forums.flightsimulator.com/t/thrustmaster-tca-quadrant-still-not-completely-working/309593/75?u=saschi13)

**Comment by Inaxair** [comment]
I'll test it

**Comment by ChruutvoLuzi** [comment]
> Quick note for the tester(s). The default bindings for the switches are quite buggy (eng. switch 1 activates both swiches ingame, eng. switch 2 makes ingame switch 2 go crazy). And the engine mode knob is not functional. I use the mapping from these posts: [LINK](https://forums.flightsimulator.com/t/thrustmaster-tca-quadrant-still-not-completely-working/309593/75?u=saschi13)

Thanks for mentioning, i was already wondering why my plane was going full [spooky](https://media.discordapp.net/attachments/314015589595086848/773227505749852180/Microsoft_Flight_Simulator_2020.11.03_-_17.47.15.01.mp4) I will test it again today 👏

**Comment by Saschl** [comment]
> > Quick note for the tester(s). The default bindings for the switches are quite buggy (eng. switch 1 activates both swiches ingame, eng. switch 2 makes ingame switch 2 go crazy). And the engine mode knob is not functional. I use the mapping from these posts: [LINK](https://forums.flightsimulator.com/t/thrustmaster-tca-quadrant-still-not-completely-working/309593/75?u=saschi13)
> 
> Thanks for mentioning, i was already wondering why my plane was going full [spooky](https://media.discordapp.net/attachments/314015589595086848/773227505749852180/Microsoft_Flight_Simulator_2020.11.03_-_17.47.15.01.mp4) I will test it again today 👏

That's really spoopy 💀  too bad Halloween is over. It's also spooky that the flaps and speed brakes are triggered with the default mapping when going to the menu. But I think that's up to MS/Asobo to fix the defaults :)

**Comment by ChruutvoLuzi** [comment]
> > > Quick note for the tester(s). The default bindings for the switches are quite buggy (eng. switch 1 activates both swiches ingame, eng. switch 2 makes ingame switch 2 go crazy). And the engine mode knob is not functional. I use the mapping from these posts: [LINK](https://forums.flightsimulator.com/t/thrustmaster-tca-quadrant-still-not-completely-working/309593/75?u=saschi13)
> > 
> > 
> > Thanks for mentioning, i was already wondering why my plane was going full [spooky](https://media.discordapp.net/attachments/314015589595086848/773227505749852180/Microsoft_Flight_Simulator_2020.11.03_-_17.47.15.01.mp4) I will test it again today 👏
> 
> That's really spoopy 💀 too bad Halloween is over. It's also spooky that the flaps and speed brakes are triggered with the default mapping when going to the menu. But I think that's up to MS/Asobo to fix the defaults :)

The flaps and spoilers is because they already mapped both axis for the upcoming addon. Unbinding them fixes it

**Comment by ChruutvoLuzi** [comment]
Name: ChruutvoLuzi | SD-PB.ch#8902
Date of testing: Current date 4.11.2020
Version of the sim: 1.10.8.0
PR Tested: #1725 
Object of test: TCA Quadrant Compatibility

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Spawned C&D
2. Started everything up as usual
3. toggled Engine Mode Selector to ignition on the Quadrant 
4. started both Engines one by one with the Quadrant 

Flight notes

Startup: Starting up the engines is no problem, however if you turn on the master switches the ignition goes on as expected, but if you want to turn it off the ignition wont go off until you turn them back on and off with your mouse.

Taxi/Takeoff : n/a

Cruise: n/a

Descent/Landing : n/a

Shutdown: as described above

Issues (If Any): 
Overall Rating: Good, with a few improvements needed, if possible
Conclusions of the testing: if its a problem of the sim limitations or wrong profiling, its a good PR, if it can be fixed through this PR, i would recommend to see that before merging. I used the Profile that @Saschl linked above. Due to hardware limitations the engine selector cant be turned on to the norm mode without assigning it to another button, dont know if its possible to code that in somehow (Button 7 not and button 8 not = Ign Norm or something) but thats more of an sobo issue

good work

**Comment by MoparCharger1969** [comment]
> Flight notes
> 
> Startup: Starting up the engines is no problem, however if you turn on the master switches the ignition goes on as expected, but if you want to turn it off the ignition wont go off until you turn them back on and off with your mouse.
> 

I'm curious about this shutdown issue you mentioned. I don't get that on my system.  With the hardware mappings from the linked post earlier, I have no trouble starting the engines with the switches (and IGN/Start mode selected), and the engines also shut down fine when the switches are turned off (and IGN start mode deselected/off).  I have not had to click the on screen engine switches with my mouse to get the engines to shut down.
 
The Norm mode issue you note is a Microsoft/Asobo issue with the button mappings in the base sim.

**Comment by ChruutvoLuzi** [comment]
> 
> 
> > Flight notes
> > Startup: Starting up the engines is no problem, however if you turn on the master switches the ignition goes on as expected, but if you want to turn it off the ignition wont go off until you turn them back on and off with your mouse.
> 
> I'm curious about this shutdown issue you mentioned. I don't get that on my system. With the hardware mappings from the linked post earlier, I have no trouble starting the engines with the switches (and IGN/Start mode selected), and the engines also shut down fine when the switches are turned off (and IGN start mode deselected/off). I have not had to click the on screen engine switches with my mouse to get the engines to shut down.
> 
> The Norm mode issue you note is a Microsoft/Asobo issue with the button mappings in the base sim.
> 
> Edit: Is the format your post is in a standard for testing these things? I'd be glad to put my results in too, but do I need to put it in the same format?

weird, i could swear i used the same configuration

the template is for QA Testers (which you can see on Discord) i dont know if you are allowed to test so you better contact a developer

**Comment by MoparCharger1969** [comment]
> > > Flight notes
> > > Startup: Starting up the engines is no problem, however if you turn on the master switches the ignition goes on as expected, but if you want to turn it off the ignition wont go off until you turn them back on and off with your mouse.
> > 
> > 
> > I'm curious about this shutdown issue you mentioned. I don't get that on my system. With the hardware mappings from the linked post earlier, I have no trouble starting the engines with the switches (and IGN/Start mode selected), and the engines also shut down fine when the switches are turned off (and IGN start mode deselected/off). I have not had to click the on screen engine switches with my mouse to get the engines to shut down.
> > The Norm mode issue you note is a Microsoft/Asobo issue with the button mappings in the base sim.
> > Edit: Is the format your post is in a standard for testing these things? I'd be glad to put my results in too, but do I need to put it in the same format?
> 
> weird, i could swear i used the same configuration
> 
> the template is for QA Testers (which you can see on Discord) i dont know if you are allowed to test so you better contact a developer

Fair enough.  I was just trying to help so this gets pushed through quicker.  I'd like to be able to get back to the Master/Dev version.  As it is, I'm sitting on this variant until that happens.

**Comment by ChruutvoLuzi** [comment]
> > > > Flight notes
> 
> > > > Startup: Starting up the engines is no problem, however if you turn on the master switches the ignition goes on as expected, but if you want to turn it off the ignition wont go off until you turn them back on and off with your mouse.
> 
> > > 
> 
> > > 
> 
> > > I'm curious about this shutdown issue you mentioned. I don't get that on my system. With the hardware mappings from the linked post earlier, I have no trouble starting the engines with the switches (and IGN/Start mode selected), and the engines also shut down fine when the switches are turned off (and IGN start mode deselected/off). I have not had to click the on screen engine switches with my mouse to get the engines to shut down.
> 
> > > The Norm mode issue you note is a Microsoft/Asobo issue with the button mappings in the base sim.
> 
> > > Edit: Is the format your post is in a standard for testing these things? I'd be glad to put my results in too, but do I need to put it in the same format?
> 
> > 
> 
> > weird, i could swear i used the same configuration
> 
> > 
> 
> > the template is for QA Testers (which you can see on Discord) i dont know if you are allowed to test so you better contact a developer
> 
> 
> 
> Fair enough.  I was just trying to help so this gets pushed through quicker.  I'd like to be able to get back to the Master/Dev version.  As it is, I'm sitting on this variant until that happens.  

All good, all good, wasnt meant to be rude. There cant be enough tests :) i was in the same position like you so i would recommend asking a dev if you can write a report but just mention that you are not officially a qa tester

**Comment by Saschl** [comment]
Thanks guys for testing! As already mentioned the engine selector switch is a bit tricky as the "norm" does not appear as a button . @ChruutvoLuzi Unfortunately I don't know about a way to implement the logic to switch it back to normal automatically.. :( as for the engine switches it's the same. I tried messing around with the templates for the switches but couldn't get the starter to stop on shutdown. As it's the same behaviour as in the stock A320 I assume it's something Asobo and/or Thrustmaster have to fix. Seeing the half broken default mapping my guess is that they just couldn't finish the proper implementation for the last update. Fingers crossed for the next update... D: 

Also if someone with more experience than me has ideas regarding the issues feel free to suggest changes. Cheers

**Comment by Saschl** [comment]
Alright I think I found a feasible solution (at least until the next update either fixes or breaks this solution) :) Please test again if you want. The starter should now shut down when the switch is in "off" position. Also the button mapping can be simplified:

**Comment by ChruutvoLuzi** [comment]
@Saschl will test it again in about 6h :)

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml [hunk 1]
```diff
@@ -294,4 +294,56 @@
             <TOOLTIPID>%((A:APU PCT RPM, Percent) 0 &gt; (A:APU PCT RPM, Percent) #AVAIL_PERCENTAGE# &lt; and)%{if}TT:COCKPIT.TOOLTIPS.UNAVAILABLE%{end} %((A:APU PCT RPM, Percent) #AVAIL_PERCENTAGE# &gt;= (A:APU SWITCH, Bool) and)%{if}TT:COCKPIT.TOOLTIPS.APU_STARTER_TURN_OFF%{else}TT:COCKPIT.TOOLTIPS.APU_STARTER_TURN_ON%{end}</TOOLTIPID>
         </UseTemplate>
     </Template>
+        <Template Name="FBW_ENGINE_Switch_Master_SubTemplate">
```
Comment (lukecologne) [inline A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml:297]: In this case, copying only the subtemplate is fine, since the "main" template isn't doing much. But you should remove the "sub" from the name. Also, the indentation of the template is a bit wrong.

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml [hunk 2]
```diff
@@ -294,4 +294,56 @@
...
+		<DefaultTemplateParameters>
+			<NODE_ID>ENGINE_Switch_Master_#ID#</NODE_ID>
+			<ANIM_NAME>ENGINE_Switch_Master_#ID#</ANIM_NAME>
+			<WWISE_EVENT_1>engine_master_switch_on</WWISE_EVENT_1> 
+			<WWISE_EVENT_2>engine_master_switch_off</WWISE_EVENT_2> 
+			<PART_ID>ENGINE_Switch_Master</PART_ID>
+			<Switch Param="TYPE">
+				<Case Value="LOCKED">
```
Comment (lukecologne) [inline A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml:305]: You should be able to remove this switch statement, since the airbus master switch isn't locked.

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml [hunk 3]
```diff
@@ -294,4 +294,56 @@
...
+					<LOCK_NODE_ID>ENGINE_Switch_Master_#ID#_Lock</NODE_ID>
+					<LOCK_ANIM_NAME>ENGINE_Switch_Master_#ID#_Lock</ANIM_NAME>
+					<VAR_SCOPE>I</VAR_SCOPE> <!-- I or L -->
+				</Case>
+			</Switch>
+			<ANIMTIP_0>TT:COCKPIT.TOOLTIPS.ENGINE#ID#_MASTER_TURN_ON</ANIMTIP_0>
+			<ANIMTIP_1>TT:COCKPIT.TOOLTIPS.ENGINE#ID#_MASTER_TURN_OFF</ANIMTIP_1>
+		</DefaultTemplateParameters>
+				<Component ID="#NODE_ID#" Node="#NODE_ID#">
+					<Update Frequency="5">
```
Comment (lukecologne) [inline A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml:315]: Why did you change this update tag? in the original, this is Once="True" instead of Frequency="5"

**Review by lukecologne — CHANGES_REQUESTED** [review]

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml (hunk 3 above)
Comment (Saschl) [reply A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml:315]: The key event fired by the quadrant actually didn't execute the code of this switch and this toggles the starter after the "toggle fuel valve" key is sent.
 I agree it's a rather bad workaround but I don't see how I can trigger the switch directly as it only seems to respond to "mouse left clicks". 
I tried using a two state switch but was not sure if this would change something. 

If you have more ideas, I'd be happy to hear!

File: A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml (hunk 2 above)
Comment (Saschl) [reply A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_Misc.xml:305]: WIll remove

**Comment by MoparCharger1969** [comment]
> Alright I think I found a feasible solution (at least until the next update either fixes or breaks this solution) :) The starter should now shut down when the switch is in "off" position. Also the button mapping can be simplified:
> 
> image
> That's enough for the engines to be started.
> 

are you saying these are the ONLY commands that need to be assigned to buttons 3 and 4 now?   We can discard all other mappings:
Set engine master 1
Magneto 1 start
Set Starter 1 (hold)
Toggle Engine 1 fuel valve (admittedly I wondered if this one could be removed anyway since it was from the default asobo mappings).

**Comment by Saschl** [comment]
@MoparCharger1969 exactly. The code in the switch should take care of the rest, although the implementation is a bit lousy and relies on periodic updates and not the event itself. But I don't know better for now

**Comment by MoparCharger1969** [comment]
> @MoparCharger1969 exactly. The code in the switch should take care of the rest, although the implementation is a bit lousy and relies on periodic updates and not the event itself. But I don't know better for now

I'm curious how this would apply to the vanilla a320.  I think those other commands are necessary on the vanilla version based on the conversation in the thread on the FS forums.  
 
I wonder if it would be best to just keep things consistent with the vanilla version so that the A32Nx mod is more compatible with other hardware controllers.  Or would the periodic updates thing also work with generic controllers too.  
 
I'm just afraid that if the changes are only relevant to the TCA throttle and aren't compatible with other hardware controllers, that the changes might not be approved.  (Again, not a dev and have limited knowledge of these things).

**Comment by Saschl** [comment]
Good point actually. IMO the last changes don't affect any other behaviour as it just tweaks the defaults a bit. So the ingame switches and any additional bindings (like the default thrustmaster bindings) work just as before. 

But right if the changes are not being approved I'll remove it again so that it behaves like vanilla.

**Comment by ChruutvoLuzi** [comment]
Tested and works as it should now, imo this is function wise good to go now. Except the Engine Mode Selector but yeah thats another story

**Comment by MoparCharger1969** [comment]
Geez... how long does it take to get this approved and merged into the master?

**Review by lukecologne — APPROVED** [review]

**Comment by lukecologne** [comment]
@Saschl sorry for taking so long to take a look again, was busy with my GPWS PR.
@MoparCharger1969 Honestly, you being annoying and rude will not speed up the process. We all have IRL stuff, our own things we work on in this mod, or other projects, it can happen that a PR is forgotton sometimes.

**Comment by MoparCharger1969** [comment]
Not trying to be rude and certainly not attacking anyone.  Just genuinely curious.  The FBW team has always been very responsive and quick to get fixes implemented.  This is the first time that I’ve been personally interested in a very specific, unique fix and it happens to take a while.  @Saschl has been fantastic with working this fix out and responding to comments, including the requests to fix a couple things.  I just hate to see it sitting idle and forgotten.

**Comment by RiderKickingMarco** [comment]
> Alright I think I found a feasible solution (at least until the next update either fixes or breaks this solution) :) The starter should now shut down when the switch is in "off" position. Also the button mapping can be simplified:
> 
> image
> That's enough for the engines to be started.
> 
> Feel free to test the latest commit if you want.

Is this the only mapping I need to set? Can the others stay default?

**Comment by ChruutvoLuzi** [comment]
> > Alright I think I found a feasible solution (at least until the next update either fixes or breaks this solution) :) The starter should now shut down when the switch is in "off" position. Also the button mapping can be simplified:
> 
> > 
> 
> > image
> 
> > That's enough for the engines to be started.
> 
> > 
> 
> > Feel free to test the latest commit if you want.
> 
> 
> 
> Is this the only mapping I need to set? Can the others stay default?

These are the only buttons you will need for Button 3 and 4. You can remove all other functions binded to those buttons

**Comment by RiderKickingMarco** [comment]
> > > Alright I think I found a feasible solution (at least until the next update either fixes or breaks this solution) :) The starter should now shut down when the switch is in "off" position. Also the button mapping can be simplified:
> > 
> > 
> > > 
> > 
> > 
> > > image
> > 
> > 
> > > That's enough for the engines to be started.
> > 
> > 
> > > 
> > 
> > 
> > > Feel free to test the latest commit if you want.
> > 
> > 
> > Is this the only mapping I need to set? Can the others stay default?
> 
> These are the only buttons you will need for Button 3 and 4. You can remove all other functions binded to those buttons

Cool! Thank you for your explanation.
